### ISSUE TITLE
Feature for handling timeouts, circuit breaker and retries

### ISSUE THREAD

**Issue by mersocarlin** [zalando-incubator/perron#72]
## TL;DR

Some suggestions that might be helpful:

- Identify when a request has failed by timeout and don't rely on `error.request === null` this scenario
- Enhance the `error` object with timeout, retries and circuit breaker errors
- Get information of each retry to easily update metrics (status code, timings, etc..)
  - Option to add `onRetrySuccess` and `onRetryError` callbacks

## Initial background 

Myself and my team have been using `perron` in some of our apps for a while now.
Big kudos for the great work, we really enjoy using it 🎉.

As the time passes, we've identified a couple of common scenarios that keep appearing in our applications and I'd like to ask you for a possible recommendation in that regard. 

The common boilerplate we use is:

```ts
import { ServiceClient } from 'perron'
import serviceConfig from '/path/to/service/config'

const serviceClient = new ServiceClient({
  circuitBreaker: { ... },
  hostname: serviceConfig.host,
  filters: [
    ServiceClient.treat4xxAsError,
    ServiceClient.treat5xxAsError
  ],
  retryOptions: { ... }
})

export const fetchDataFromService = async () => {
  try {
    const response = await serviceClient.request({
      headers: { ... },
      method: 'get',
      pathname: 'path/to/api/resource',
      query: { ... },
      timeout: serviceConfig.timeout,
    })
    // track timings
    // track retries (when request succeeded by retry)
    return response
  } catch(error) {
    // track timings
    // update error metric
    // track retries (when request failed by retry)
    throw error
  }
}
```

As the codebase grows, the same pattern keeps happening across our applications:

- Instantiate the service client
- Wrap the api call in a `try/catch`
- If success
  - Track timings
  - Track retries (when there was a retry attempt)
  - Return response
- If failure
  - Track timings
  - Track error metric
  - Track retries (when there was a retry attempt)
  - Throw error

There's also a couple of metrics we need to keep track in order to ensure our service is or is not working as expected: 

- Failure by known status codes: 401, 404, 429, 500, 503, ...
- Failure by request timeout
- Failure by circuit breaker open
 - Failure by retry

## A few findings in case the request fails

At the end of every request, we send the data associated to the success or failure request to a 3rd library tool for further investigation.
It’s important we track the right metrics, for each type of failure so we don't make wrong assumptions based on an error that was thrown by our services. 

### Failure by known status codes 

In case the api request fails and throws, for example a 400, it's rather easy to identify what might have gone wrong, treat the error, save the metric and send/throw the error to whoever has dispatched the api call.

### Failure by request timeout

It would be nice to identify on our side if the api request has failed by timeout (either if the server we are making the request to is timing out or if the request is taking longer than the timeout prop defined in the request).
For that reason, we ended up changing the catch part to something similar to:

```ts
export const fetchDataFromService = async () => {
  try {
    const response = await serviceClient.request({ ... })
    // track timings
    // track retries (when request succeeded by retry)
    return response
  } catch(error) {
    error.metadata = {
      errorByTimeout: error.response === null,
    }
    // track timings
    // update error metric
    // track retries (when request failed by retry)
    throw error
  }
}
```

### Failure by circuit breaker open

As already mentioned, circuit breaker is also an important metric we’d like to keep an eye on, especially if the request fails. For this reason, the code changed to: 

```ts
export const fetchDataFromService = async () => {
  try {
    const response = await serviceClient.request({ ... })
    // track timings
    // track retries (when request succeeded by retry)
    return response
  } catch(error) {
    error.metadata = {
      errorByTimeout: error.response === null,
      errorByCircuitBreakerOpen: err.type === ServiceClient.CIRCUIT_OPEN,
    }
    // track timings
    // update error metric
    // track retries (when request failed by retry)
    throw error
  }
}
```

### Failure by retry

Another important metric we need to keep an eye on is whether the request has failed/succeeded by retry. We couldn’t find a proper way to track that as `perron` does not provide, for example, `onRetrySuccess` or `onRetryError` callbacks.
It's true that `retryOptions` is very handy for handling retries, but we are not able to identify, for example, whether a successful request has finished straight away or if there was a retry process (1, 2, 3, N attempts) before we get the result.
For the time being, the current solution lives in the service client creation so later we can also track this metric once the request finishes:

```ts
import { ServiceClient } from 'perron'
import { get } from 'lodash'
import serviceConfig from '/path/to/service/config'

const serviceClient = new ServiceClient({
  ...
  retryOptions: {
    ...
    shouldRetry: (err, request) => {
      let shouldRetry = false
      // custom login for handling retry
      return shouldRetry
    },
    onRetry: (currentAttempt, err, request) => {
      // store the number of attemps a retry has been dispatched
      request.metadata = {
        retryAttempts: currentAttempt,
      }
    },
  }
})

export const fetchDataFromService = async () => {
  try {
    const response = await serviceClient.request({ ... })
    // track timings
                
    const retryAttempts = get(error, 'request.metadata.retryAttempts', 0)
    if (retryAttempts > 0) {
      // track retries
    }
    
    return response
  } catch(error) {
    // track timings
    // update error metric
      
    const retryAttempts = get(error, 'request.metadata.retryAttempts', 0)
    if (retryAttempts > 0) {
      // track retries
    }
    
    throw error
  }
}
```

## Help wanted

Although this solution might work for now, I'm not entirely comfortable in making mutations in the `error` and `request` objects thrown by `perron`, neither by relying on `error.response === null` for situation A, `error.response === undefined` for situation B or `error.type` for situation C.
Circuit breaker is an example where the error type is defined in the error object, but it’s not something we could rely on in case it fails by timeout.

As I said, it works now, but it can easily break in case `perron` changes the API in the future.

Do you think there’s a better way to identify/handle such error types? 
Would `err.metadata` and `response.metadata` (or any type of extra info) be something worth taking into consideration?

Is that ok if we open a PR addressing these scenarios?
I know it's a lot to change in one go, so I'm happy to contribute in smaller chunks if necessary.

Sorry for the big issue description. I couldn't find a proper way to demonstrate everything by not using code samples :)

**Comment by mersocarlin** [comment]
@grassator @shuhei 

Just tagging both of you here :) 
I also don't have access to add labels to this issue.

**Comment by grassator** [comment]
@mersocarlin Thank you for reaching and it is great to see perron being useful for people.

In general I like the idea of having more meta information, I'm just wondering what would be the best interface for doing something like this. Does your setup require metrics to be reported right at the time of the retry, or can it be at the time of the final response / error?

**Comment by mersocarlin** [comment]
Hey @grassator, thanks for you reply! 

I think it's important to know how many times a retry has happened rather than when they happened.
If I had to choose, I'd say that I could use it when the final response/failure comes in, specially because I'll have the entire logic in a single place.

I'm kind of doing this right now by setting custom metadata on `onRetry` method to later process the outcome of the request.

**Comment by grassator** [comment]
After thinking about this for bit I can see several ways we could implement this:

1. Add more events / callbacks to the client instance that would receive the request instance along with information on what happened. You would then be able to correlate the event with a particular request through a Map or WeakMap.
1. Introduce “request” object that would be a subclass of Promise that would keep backwards compatibility but allow to subscribe to extra events.
1. Provide a “log” or “trace” field on the response that would be a list of events that led to that result in order.

I’m honestly of sure what is the best option here and would be glad to hear opinions on this. @shuhei @DeTeam

**Comment by shuhei** [comment]
@mersocarlin Thanks for collecting issues and writing up! Here are my two cents:

## Error reason

I think an error always has only one reason. Because of that, I prefer to have a single field on error object rather than having multiple metadata fields such as `errorByTimeout` and `errorByCircuitBreakerOpen`. [Node.js error code](https://nodejs.org/api/errors.html#errors_error_code) seems to be a good example. We can put more detailed reason from `req.ts` to `ServiceClientError.type` in `client.ts`, or add a new field if we want to avoid a breaking change.

## Timings

I am curious how fine-grained timings you need. I added the timings support, but I realized after some time that I didn't need detailed timings per request. Currently I monitor `lookup` function's latency globally and the entire time from request start to response end per request type.

## Retries and other metadata

I don't have a strong opinion here, but @grassator's option 3 looks straightforward to me.

**Comment by grassator** [comment]
I've started working on a variation of option 3 that seems promising. Hopefully will have something on the weekend.

**Comment by grassator** [comment]
@mersocarlin Please take a look at https://github.com/zalando-incubator/perron/pull/76 - I believe it should cover all your needs.

**Comment by mersocarlin** [comment]
@grassator 

Many thanks for your help and support.
I had a look a the code and enhancing the error types seems a good approach. 

Once the code is merged, I'm happy to give it a try and provide extra feedback if needed :)

**Comment by grassator** [comment]
@mersocarlin I've released 0.9.0 with the changes. Once you have a chance to try please let us know if you have any additional feedback or we can close this issue.

**Comment by mersocarlin** [comment]
hey @grassator 

I had a chance to test v0.9.1 out and it looks great! 
It's so much easier to assert the errors now with `RequestConnectionTimeoutError`, `CircuitOpenError` or `MaximumRetriesReachedError` 🎉 🎉 

The hacky `metadata` stuff can go away now with the new error types.

Feel free to close this issue btw.

Sorry for the delay, I wanted to dedicate a fair amount of time trying it out before replying you :) 

I might have one or two feedbacks here, but I can open another issue, not worries about that. 

Many thanks for now, I really appreciate it o/

**Comment by grassator** [comment]
Great to hear and looking forward to extra feedback.
